File: swift/Mouse/Runtimes.swift
import Foundation
import CryptoKit

/// Language runtimes that arrive as WebAssembly.
///
/// The rule this file exists to keep: a runtime is DOWNLOADED, never bundled. Shipping CPython
/// inside the app would add 45 MB to every install for a language most projects never touch, and
/// would freeze its version to whenever the app was last released. `pkg install python` fetches
/// the official wasm32-wasi build, checks it against a recorded hash, and unpacks it into the
/// app's Application Support directory — not Caches, because iOS may evict Caches under pressure
/// and a language silently disappearing mid-project is worse than the disk it costs.
///
/// What runs the artifact is the phase-G engine's own `node:wasi`: every one of python.wasm's 42
/// imports is `wasi_snapshot_preview1`, and every one of those is already implemented here.
enum RuntimeCatalog {
    struct Entry {
        let name: String
        let version: String
        let url: URL
        /// The download's size and SHA-256. The size is what the install line quotes before it
        /// starts; the hash is what makes the install refuse a corrupted or substituted archive.
        let downloadBytes: Int
        let sha256: String
        /// The interpreter inside the archive, and the directory its standard library lives in.
        let wasm: String
        let library: String?
        /// A one-line description of what the runtime is, for `pkg list`.
        let summary: String
    }

    static let python = Entry(
        name: "python",
        version: "3.14.6",
        url: URL(string: "https://github.com/brettcannon/cpython-wasi-build/releases/download/v3.14.6/python-3.14.6-wasi_sdk-24.zip")!,
        downloadBytes: 14_231_464,
        sha256: "73bf2e9774c4d8820d0877ec5db0b963df3a9611fc2a63838aeaee29dfd034e6",
        wasm: "python.wasm",
        library: "lib",
        summary: "CPython, the official wasm32-wasi build"
    )

    static let all: [Entry] = [python]

    static func entry(named name: String) -> Entry? {
        all.first { $0.name == name }
    }
}

/// Where installed runtimes live, and how they get there.
enum RuntimeStore {
    static let root: URL = {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        let directory = base.appendingPathComponent("MouseRuntimes", isDirectory: true)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory
    }()

    struct Installed {
        let name: String
        let version: String
        let directory: URL
        let wasm: URL
        let library: URL?
    }

    static func installed(_ name: String) -> Installed? {
        guard let entry = RuntimeCatalog.entry(named: name) else { return nil }
        let directory = root.appendingPathComponent(entry.name, isDirectory: true)
        let wasm = directory.appendingPathComponent(entry.wasm)
        guard FileManager.default.fileExists(atPath: wasm.path) else { return nil }
        let library = entry.library.map { directory.appendingPathComponent($0, isDirectory: true) }
        return Installed(name: entry.name, version: entry.version, directory: directory,
                         wasm: wasm, library: library)
    }

    static func installedNames() -> [String] {
        RuntimeCatalog.all.compactMap { installed($0.name) != nil ? $0.name : nil }
    }

    enum InstallError: Error, CustomStringConvertible {
        case download(String)
        case integrity(expected: String, got: String)
        case unpack(String)
        case missingInterpreter(String)

        var description: String {
            switch self {
            case .download(let why): "download failed: \(why)"
            // Naming both hashes matters: it separates "the network mangled it" from "this is not
            // the archive we recorded", and only the second is alarming.
            case .integrity(let expected, let got):
                "the download does not match its recorded hash\n  expected \(expected)\n  got      \(got)"
            case .unpack(let why): "could not unpack the archive: \(why)"
            case .missingInterpreter(let name): "the archive did not contain \(name)"
            }
        }
    }

    /// Download, verify, unpack, reporting progress as it goes — a multi-megabyte download with
    /// no output is indistinguishable from a hang.
    ///
    /// Progress arrives as a STREAM rather than through a callback the caller hands in: the work
    /// runs off the main actor (a 30 MB inflate has no business on the UI thread) while the
    /// caller consumes each line on whatever actor it lives on, so no closure has to cross
    /// between them. A `(String) -> Void` parameter here does not compile under strict
    /// concurrency, and the honest reason is that it would genuinely be a data race.
    static func install(_ entry: RuntimeCatalog.Entry) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream { continuation in
            let work = Task.detached {
                do {
                    try await performInstall(entry) { continuation.yield($0) }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in work.cancel() }
        }
    }

    private static func performInstall(_ entry: RuntimeCatalog.Entry,
                                       note: @Sendable (String) -> Void) async throws {
        note("fetching \(entry.name) \(entry.version) (\(entry.downloadBytes / 1_000_000) MB)")
        let data: Data
        do {
            let (downloaded, response) = try await URLSession.shared.data(from: entry.url)
            if let http = response as? HTTPURLResponse, http.statusCode != 200 {
                throw InstallError.download("HTTP \(http.statusCode)")
            }
            data = downloaded
        } catch let error as InstallError {
            throw error
        } catch {
            throw InstallError.download(error.localizedDescription)
        }

        let digest = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
        guard digest == entry.sha256 else {
            throw InstallError.integrity(expected: entry.sha256, got: digest)
        }

        let directory = root.appendingPathComponent(entry.name, isDirectory: true)
        // A half-written runtime is worse than none: unpack beside the target and swap, so an
        // interrupted install leaves the previous one intact.
        let staging = root.appendingPathComponent(".\(entry.name).incoming", isDirectory: true)
        try? FileManager.default.removeItem(at: staging)
        try FileManager.default.createDirectory(at: staging, withIntermediateDirectories: true)
        do {
            try ZipArchive.extract(data, to: staging)
        } catch {
            try? FileManager.default.removeItem(at: staging)
            throw InstallError.unpack("\(error)")
        }
        guard FileManager.default.fileExists(atPath: staging.appendingPathComponent(entry.wasm).path) else {
            try? FileManager.default.removeItem(at: staging)
            throw InstallError.missingInterpreter(entry.wasm)
        }
        try? FileManager.default.removeItem(at: directory)
        try FileManager.default.moveItem(at: staging, to: directory)
        note("installed \(entry.name) \(entry.version)")
    }

    static func remove(_ name: String) -> Bool {
        guard let entry = RuntimeCatalog.entry(named: name) else { return false }
        let directory = root.appendingPathComponent(entry.name, isDirectory: true)
        guard FileManager.default.fileExists(atPath: directory.path) else { return false }
        try? FileManager.default.removeItem(at: directory)
        return true
    }
}

/// A reader for the ZIP format, because iOS has no `unzip` and the runtimes upstream publishes
/// are zips. Only what an unpack needs: walk the central directory, then copy stored entries and
/// inflate deflated ones. The inflate itself is the one `TarGz` already uses for npm
/// tarballs — Apple's `compression` framework in raw-DEFLATE mode, which is exactly what a zip
/// entry holds.
enum ZipArchive {
    struct FormatError: Error, CustomStringConvertible {
        let description: String
        init(_ description: String) { self.description = description }
    }

    static func extract(_ data: Data, to destination: URL) throws {
        let bytes = [UInt8](data)
        guard let eocd = endOfCentralDirectory(bytes) else {
            throw FormatError("no end-of-central-directory record — not a zip")
        }
        var offset = eocd.centralDirectoryOffset
        let manager = FileManager.default

        for _ in 0..<eocd.entryCount {
            guard offset + 46 <= bytes.count,
                  read32(bytes, offset) == 0x0201_4b50 else {
                throw FormatError("central directory entry is malformed")
            }
            let method = Int(read16(bytes, offset + 10))
            let compressedSize = Int(read32(bytes, offset + 20))
            let uncompressedSize = Int(read32(bytes, offset + 24))
            let nameLength = Int(read16(bytes, offset + 28))
            let extraLength = Int(read16(bytes, offset + 30))
            let commentLength = Int(read16(bytes, offset + 32))
            let localOffset = Int(read32(bytes, offset + 42))
            guard offset + 46 + nameLength <= bytes.count else {
                throw FormatError("entry name runs past the end of the archive")
            }
            let name = String(decoding: bytes[(offset + 46)..<(offset + 46 + nameLength)], as: UTF8.self)
            offset += 46 + nameLength + extraLength + commentLength

            // An entry naming `..` or an absolute path can write anywhere on disk. Archives are
            // downloads; refusing this is not paranoia, it is the whole reason to parse rather
            // than shell out.
            let components = name.split(separator: "/", omittingEmptySubsequences: true).map(String.init)
            guard !name.hasPrefix("/"), !components.contains("..") else {
                throw FormatError("archive entry escapes its directory: \(name)")
            }
            let target = components.reduce(destination) { $0.appendingPathComponent($1) }
            if name.hasSuffix("/") {
                try manager.createDirectory(at: target, withIntermediateDirectories: true)
                continue
            }

            guard localOffset + 30 <= bytes.count, read32(bytes, localOffset) == 0x0403_4b50 else {
                throw FormatError("local header for \(name) is malformed")
            }
            // The local header repeats the name and extra-field lengths, and its own extra field
            // is frequently a different length from the central directory's — reading the
            // central one here lands in the middle of the data.
            let localNameLength = Int(read16(bytes, localOffset + 26))
            let localExtraLength = Int(read16(bytes, localOffset + 28))
            let start = localOffset + 30 + localNameLength + localExtraLength
            guard start + compressedSize <= bytes.count else {
                throw FormatError("data for \(name) runs past the end of the archive")
            }
            let payload = data.subdata(in: start..<(start + compressedSize))

            let contents: Data
            switch method {
            case 0: contents = payload
            case 8: contents = try TarGz.inflateRaw(payload)
            default: throw FormatError("\(name) uses compression method \(method), which is not deflate or stored")
            }
            guard contents.count == uncompressedSize else {
                throw FormatError("\(name) unpacked to \(contents.count) bytes, not the \(uncompressedSize) it declares")
            }
            try manager.createDirectory(at: target.deletingLastPathComponent(),
                                        withIntermediateDirectories: true)
            try contents.write(to: target)
        }
    }

    private struct EndRecord {
        let entryCount: Int
        let centralDirectoryOffset: Int
    }

    /// The EOCD sits at the end, after a comment of unknown length, so it is found by scanning
    /// backwards for its signature.
    private static func endOfCentralDirectory(_ bytes: [UInt8]) -> EndRecord? {
        guard bytes.count >= 22 else { return nil }
        var index = bytes.count - 22
        let floor = max(0, bytes.count - 22 - 0xffff)
        while index >= floor {
            if read32(bytes, index) == 0x0605_4b50 {
                return EndRecord(entryCount: Int(read16(bytes, index + 10)),
                                 centralDirectoryOffset: Int(read32(bytes, index + 16)))
            }
            index -= 1
        }
        return nil
    }

    private static func read16(_ bytes: [UInt8], _ at: Int) -> UInt16 {
        guard at + 2 <= bytes.count else { return 0 }
        return UInt16(bytes[at]) | (UInt16(bytes[at + 1]) << 8)
    }

    private static func read32(_ bytes: [UInt8], _ at: Int) -> UInt32 {
        guard at + 4 <= bytes.count else { return 0 }
        return UInt32(bytes[at]) | (UInt32(bytes[at + 1]) << 8)
            | (UInt32(bytes[at + 2]) << 16) | (UInt32(bytes[at + 3]) << 24)
    }
}

File: kotlin/packages/src/main/kotlin/com/reagentsystems/mouse/packages/Runtimes.kt
package com.reagentsystems.mouse.packages

import java.io.File
import java.net.HttpURLConnection
import java.net.URI
import java.security.MessageDigest

/**
 * Language runtimes that arrive as WebAssembly. A port of `swift/Mouse/Runtimes.swift`, reading
 * the SAME `swift/Runtimes.json` — the catalog is data, and one file for both platforms is the
 * only way the parity claim stays falsifiable.
 *
 * The rule this file exists to keep: a runtime is DOWNLOADED, never bundled. Shipping CPython
 * inside the APK would add ~45 MB to every install for a language most projects never touch, and
 * would freeze its version to whenever the app was last released. `pkg install python` fetches
 * the official wasm32-wasi build, checks it against a recorded hash, and unpacks it into the
 * app's files directory — never the cache, because Android may evict caches under pressure and a
 * language silently disappearing mid-project is worse than the disk it costs.
 *
 * On Android this rule is not merely a preference, it is the only door: since API 29 an app
 * targeting 29+ cannot exec a binary out of its own writable data directory, and this app targets
 * 35. Downloaded NATIVE binaries are impossible; downloaded wasm is not, because nothing execs it
 * — the phase-G engine interprets it.
 *
 * What runs the artifact is that engine's `node:wasi`, which is milestone 3. This file is the
 * half that can land before it: fetch, verify, unpack, list, remove.
 */
object RuntimeCatalog {
    enum class Archive(val id: String) {
        ZIP("zip"),
        TAR_GZ("tar.gz"),
        ;

        companion object {
            fun of(id: String): Archive? = entries.firstOrNull { it.id == id }
        }
    }

    data class Entry(
        val name: String,
        val version: String,
        val url: String,
        /**
         * The download's size and SHA-256. The size is what the install line quotes before it
         * starts; the hash is what makes the install refuse a corrupted or substituted archive.
         */
        val downloadBytes: Long,
        val sha256: String,
        /**
         * How the download unpacks, and how many leading path components to drop while it does
         * (release tarballs wrap their content in a versioned directory; the zips here are flat).
         */
        val archive: Archive,
        val strip: Int?,
        /**
         * The interpreter inside the unpacked archive, and the directory its standard library
         * lives in — both relative to the runtime's install directory.
         */
        val wasm: String,
        val library: String?,
        /**
         * The command names this runtime answers to at the prompt (`python` and `python3` are one
         * interpreter). The catalog knows them for UNINSTALLED runtimes too, which is what lets
         * the shell answer "not installed" instead of "command not found".
         */
        val commands: List<String>,
        /**
         * The interpreter's environment. `{root}` expands to the runtime's mount path
         * (`/usr/lib/<name>`) at launch, so an entry can name its own stdlib without knowing where
         * installs live.
         */
        val env: Map<String, String>,
        /**
         * Whether bare script arguments are rewritten to absolute virtual paths before launch.
         * WASI interpreters have no cwd of their own, so `python hello.py` only works if the shell
         * turns `hello.py` into `/hello.py` — flags and inline code stay untouched.
         */
        val rewriteScriptPaths: Boolean,
        /** A one-line description of what the runtime is, for `pkg list`. */
        val summary: String,
    )

    class CatalogError(message: String) : Exception(message)

    /**
     * Parse the catalog. A malformed entry THROWS rather than being skipped: a silently dropped
     * runtime shows up as "command not found" for a language the catalog plainly lists, which is
     * the least debuggable failure this layer can produce.
     */
    fun parse(text: String): List<Entry> {
        val root = Json.parseObjectOrNull(text) ?: throw CatalogError("the catalog is not a JSON object")
        val runtimes = root["runtimes"] as? List<*> ?: throw CatalogError("the catalog has no `runtimes` array")
        return runtimes.map { element ->
            val entry = element.asObject() ?: throw CatalogError("a catalog entry is not an object")
            fun string(key: String): String =
                entry[key].asString() ?: throw CatalogError("a catalog entry has no `$key`")
            val name = string("name")
            val archiveId = string("archive")
            Entry(
                name = name,
                version = string("version"),
                url = string("url"),
                downloadBytes = (entry["downloadBytes"] as? Number)?.toLong()
                    ?: throw CatalogError("$name has no `downloadBytes`"),
                sha256 = string("sha256"),
                archive = Archive.of(archiveId)
                    ?: throw CatalogError("$name declares archive `$archiveId`, which is not zip or tar.gz"),
                strip = (entry["strip"] as? Number)?.toInt(),
                wasm = string("wasm"),
                library = entry["library"].asString(),
                commands = (entry["commands"] as? List<*>)?.mapNotNull { it.asString() }
                    ?: throw CatalogError("$name has no `commands`"),
                env = entry["env"].asStringMap(),
                rewriteScriptPaths = entry["rewriteScriptPaths"] as? Boolean ?: false,
                summary = string("summary"),
            )
        }
    }

    fun entry(entries: List<Entry>, name: String): Entry? = entries.firstOrNull { it.name == name }

    /** The entry that answers to a typed command — installed or not. */
    fun forCommand(entries: List<Entry>, command: String): Entry? =
        entries.firstOrNull { command in it.commands }
}

/**
 * Where installed runtimes live, and how they get there.
 *
 * Unlike the iOS original this takes its base directory as a parameter instead of reaching for
 * one: `:packages` is a pure-JVM module with no `android.*` on its classpath (that is what makes
 * it gatable without an emulator), so it cannot ask a `Context` for `filesDir`. The app passes it
 * in; the harness passes a temp directory.
 */
class RuntimeStore(base: File) {
    /**
     * On disk the store is `MouseRuntimes/usr/lib/<name>`, and the shell mounts the whole `usr`
     * directory at `/usr`. The extra two levels are not decoration: an interpreter that
     * canonicalizes its load paths (realpath) walks `/usr`, then `/usr/lib`, then
     * `/usr/lib/<name>` component by component, and a component that exists only as a mount-table
     * prefix answers ENOENT. With the layout mirrored on disk, every ancestor a runtime names is
     * a real directory.
     */
    val usr: File = File(base, "MouseRuntimes/usr")
    val root: File = File(usr, "lib")

    init {
        root.mkdirs()
    }

    data class Installed(
        val name: String,
        val version: String,
        val directory: File,
        val wasm: File,
        val library: File?,
    )

    fun installed(entry: RuntimeCatalog.Entry): Installed? {
        val directory = File(root, entry.name)
        val wasm = File(directory, entry.wasm)
        if (!wasm.isFile) return null
        return Installed(
            name = entry.name,
            version = entry.version,
            directory = directory,
            wasm = wasm,
            library = entry.library?.let { File(directory, it) },
        )
    }

    fun installedNames(entries: List<RuntimeCatalog.Entry>): List<String> =
        entries.filter { installed(it) != null }.map { it.name }

    sealed class InstallError(message: String) : Exception(message) {
        class Download(why: String) : InstallError("download failed: $why")

        /**
         * Naming both hashes matters: it separates "the network mangled it" from "this is not the
         * archive we recorded", and only the second is alarming.
         */
        class Integrity(expected: String, got: String) : InstallError(
            "the download does not match its recorded hash\n  expected $expected\n  got      $got",
        )

        class Unpack(why: String) : InstallError("could not unpack the archive: $why")
        class MissingInterpreter(name: String) : InstallError("the archive did not contain $name")
    }

    /**
     * Download, verify, unpack, reporting progress as it goes — a multi-megabyte download with no
     * output is indistinguishable from a hang. Runs wherever the caller puts it; the caller is
     * responsible for keeping it off the main thread (a 30 MB inflate has no business on the UI
     * thread), which on iOS the compiler enforces and here it cannot.
     */
    fun install(
        entry: RuntimeCatalog.Entry,
        note: (String) -> Unit = {},
        fetch: (String) -> ByteArray = ::download,
    ) {
        note("fetching ${entry.name} ${entry.version} (${entry.downloadBytes / 1_000_000} MB)")
        val data = try {
            fetch(entry.url)
        } catch (e: InstallError) {
            throw e
        } catch (e: Exception) {
            throw InstallError.Download(e.message ?: e.toString())
        }

        val digest = MessageDigest.getInstance("SHA-256").digest(data)
            .joinToString("") { "%02x".format(it) }
        if (digest != entry.sha256) throw InstallError.Integrity(entry.sha256, digest)

        val directory = File(root, entry.name)
        // A half-written runtime is worse than none: unpack beside the target and swap, so an
        // interrupted install leaves the previous one intact.
        val staging = File(root, ".${entry.name}.incoming")
        staging.deleteRecursively()
        staging.mkdirs()
        try {
            // Both unpackers refuse entries that escape the destination (`..`, absolute names) —
            // an archive is a download, and a download must not write outside its directory.
            when (entry.archive) {
                RuntimeCatalog.Archive.ZIP -> ZipArchive.extract(data, staging)
                RuntimeCatalog.Archive.TAR_GZ -> TarGz.extract(data, staging, entry.strip ?: 0)
            }
        } catch (e: Exception) {
            staging.deleteRecursively()
            throw InstallError.Unpack(e.message ?: e.toString())
        }
        if (!File(staging, entry.wasm).isFile) {
            staging.deleteRecursively()
            throw InstallError.MissingInterpreter(entry.wasm)
        }
        directory.deleteRecursively()
        if (!staging.renameTo(directory)) {
            staging.deleteRecursively()
            throw InstallError.Unpack("could not move the unpacked runtime into place")
        }
        note("installed ${entry.name} ${entry.version}")
    }

    fun remove(entry: RuntimeCatalog.Entry): Boolean {
        val directory = File(root, entry.name)
        if (!directory.isDirectory) return false
        return directory.deleteRecursively()
    }

    companion object {
        /**
         * The default transport, and public so the gate can drive it directly: a harness that
         * substituted its own download would leave the redirect handling below ungated, and that
         * is precisely the part with a failure mode.
         *
         * Redirects are followed BY HAND. `HttpURLConnection` refuses to follow one that changes
         * protocol, and every release artifact in the catalog is a GitHub URL that redirects to a
         * separate object host — so an automatic follow can land on a 302 body, and the hash check
         * would then report a mismatch for something that was never the archive.
         */
        fun download(url: String): ByteArray {
            var target = url
            for (hop in 0 until 5) {
                val connection = (URI(target).toURL().openConnection() as HttpURLConnection).apply {
                    requestMethod = "GET"
                    connectTimeout = 30_000
                    readTimeout = 120_000
                    instanceFollowRedirects = false
                }
                try {
                    val status = connection.responseCode
                    if (status in 300..399) {
                        target = connection.getHeaderField("Location")
                            ?: throw InstallError.Download("HTTP $status with no Location header")
                        continue
                    }
                    if (status !in 200..299) throw InstallError.Download("HTTP $status")
                    return connection.inputStream.use { it.readBytes() }
                } finally {
                    connection.disconnect()
                }
            }
            throw InstallError.Download("too many redirects")
        }
    }
}

/**
 * A reader for the ZIP format, because Android's framework unzip is `java.util.zip`, whose
 * `ZipInputStream` cannot be trusted with an archive that has no usable local headers — and,
 * more to the point, because an archive is a DOWNLOAD and refusing entries that escape their
 * directory is the whole reason to parse rather than hand it to something that will not.
 *
 * Only what an unpack needs: walk the central directory, then copy stored entries and inflate
 * deflated ones. The inflate is `TarGz.inflateRaw`, the same raw-DEFLATE path npm tarballs use,
 * which is exactly what a zip entry holds. A port of `ZipArchive` in `swift/Mouse/Runtimes.swift`.
 */
object ZipArchive {
    class FormatError(message: String) : Exception(message)

    fun extract(data: ByteArray, destination: File) {
        val end = endOfCentralDirectory(data)
            ?: throw FormatError("no end-of-central-directory record — not a zip")
        var offset = end.centralDirectoryOffset

        for (index in 0 until end.entryCount) {
            if (offset + 46 > data.size || read32(data, offset) != 0x02014b50L) {
                throw FormatError("central directory entry is malformed")
            }
            val method = read16(data, offset + 10)
            val compressedSize = read32(data, offset + 20).toInt()
            val uncompressedSize = read32(data, offset + 24).toInt()
            val nameLength = read16(data, offset + 28)
            val extraLength = read16(data, offset + 30)
            val commentLength = read16(data, offset + 32)
            val localOffset = read32(data, offset + 42).toInt()
            if (offset + 46 + nameLength > data.size) {
                throw FormatError("entry name runs past the end of the archive")
            }
            val name = String(data, offset + 46, nameLength, Charsets.UTF_8)
            offset += 46 + nameLength + extraLength + commentLength

            // An entry naming `..` or an absolute path can write anywhere on disk.
            val components = name.split("/").filter { it.isNotEmpty() }
            if (name.startsWith("/") || components.contains("..")) {
                throw FormatError("archive entry escapes its directory: $name")
            }
            val target = components.fold(destination) { parent, component -> File(parent, component) }
            if (name.endsWith("/")) {
                target.mkdirs()
                continue
            }

            if (localOffset + 30 > data.size || read32(data, localOffset) != 0x04034b50L) {
                throw FormatError("local header for $name is malformed")
            }
            // The local header repeats the name and extra-field lengths, and its own extra field
            // is frequently a different length from the central directory's — reading the central
            // one here lands in the middle of the data.
            val localNameLength = read16(data, localOffset + 26)
            val localExtraLength = read16(data, localOffset + 28)
            val start = localOffset + 30 + localNameLength + localExtraLength
            if (start + compressedSize > data.size) {
                throw FormatError("data for $name runs past the end of the archive")
            }
            val payload = data.copyOfRange(start, start + compressedSize)

            val contents = when (method) {
                0 -> payload
                8 -> TarGz.inflateRaw(payload)
                else -> throw FormatError(
                    "$name uses compression method $method, which is not deflate or stored",
                )
            }
            if (contents.size != uncompressedSize) {
                throw FormatError(
                    "$name unpacked to ${contents.size} bytes, not the $uncompressedSize it declares",
                )
            }
            target.parentFile?.mkdirs()
            target.writeBytes(contents)
        }
    }

    private class EndRecord(val entryCount: Int, val centralDirectoryOffset: Int)

    /**
     * The EOCD sits at the end, after a comment of unknown length, so it is found by scanning
     * backwards for its signature.
     */
    private fun endOfCentralDirectory(data: ByteArray): EndRecord? {
        if (data.size < 22) return null
        var index = data.size - 22
        val floor = maxOf(0, data.size - 22 - 0xffff)
        while (index >= floor) {
            if (read32(data, index) == 0x06054b50L) {
                return EndRecord(read16(data, index + 10), read32(data, index + 16).toInt())
            }
            index -= 1
        }
        return null
    }

    private fun read16(data: ByteArray, at: Int): Int {
        if (at + 2 > data.size) return 0
        return (data[at].toInt() and 0xFF) or ((data[at + 1].toInt() and 0xFF) shl 8)
    }

    /** Unsigned: a zip offset past 2 GB is still a valid `Int` here only because we check it. */
    private fun read32(data: ByteArray, at: Int): Long {
        if (at + 4 > data.size) return 0
        return (data[at].toLong() and 0xFF) or
            ((data[at + 1].toLong() and 0xFF) shl 8) or
            ((data[at + 2].toLong() and 0xFF) shl 16) or
            ((data[at + 3].toLong() and 0xFF) shl 24)
    }
}
